Refactor saveloading ui

jmoerschbach opened the pull request:
Hi,

in #3561 I promised to revisit this code again. I hav done so ;)

(Hopefully) no semantic changes at all, only making the code's intention clearer...

What I did:

- Further separating of concerns: Extracted savegame loading code (SavegameLoader) and savegame deleting code (SavegameDeleter) into own classes
- Introduced small inheritance hierachy for both of these classes to adhere to different loading/deleting of replays/multiplayer/singleplayer

I think there is still room for improvement, for example I do not like the exposure of "internal" objects (via LoadOrSaveGame#table(), #table_box(), #game_details(), #delete_button())

Can someone explain or point me to documentation about how replays are created/saved? What is GameController::GameType::kReplay actually used for? In the old code (and currently in this pr, too) it is not evaluated, is_valid_game always returns true, but I think this might be wrong in the first place?

Feedback very much appreciated :)

Thanks

Jonas

jmoerschbach committed:
extracted methods

jmoerschbach committed:
pass entire list via reference instead of passing list size only

jmoerschbach committed:
Merge branch 'master' into refactor-saveloading-ui

jmoerschbach committed:
introducing SavegameDeleter and extracting SavegameData into separate file

jmoerschbach committed:
use savegamedeleter to encapsulate all the deletion code

jmoerschbach committed:
moved some more code

jmoerschbach committed:
introducing savegameloader, adding virtual destructor

jmoerschbach committed:
add to CMakeLists.txt

jmoerschbach committed:
introducing MultiPlayerLoader, adding logs

jmoerschbach committed:
better distinction between savegame loading from file vs directory

jmoerschbach committed:
Merge branch 'master' into refactor-saveloading-ui

rename functions, removed unnecessary includes

jmoerschbach committed:
extracted some methods & do not attempt to show game details of a directory

jmoerschbach committed:
fix compile error, remove unnecessary include

jmoerschbach committed:
minor refactorings

jmoerschbach committed:
Merge branch 'master' into refactor-saveloading-ui

jmoerschbach committed:
switch instead of if / fixed includes / SinglePlayerLoader

jmoerschbach committed:
Merge branch 'master' of https://github.com/widelands/widelands into refactor-saveloading-ui

jmoerschbach committed:
added some comments / moved code / deleted commented code

jmoerschbach commented:
Holy shit, what happened here with the code formatting? Oo
Most of the files I didn't even opened in my editor... How come the format has changed?
Is fixing this somehow automated? Do I need to do something about it?

Noordfrees commented:
Perhaps you have the deprecated clang-format-githook installed on your machine?

Please go to https://github.com/jmoerschbach/widelands/runs/493304990 and click "Re-run jobs", that should fix the formatting.

actions-user committed:
61 files were automatically formatted.

jmoerschbach commented:
Thanks @Noordfrees! Looks better now :)

Will check my git hook tomorrow...

jmoerschbach committed:
fix includes and other codecheck warnings

actions-user committed:
'src/wui/savegameloader.cc' was automatically formatted.

jmoerschbach commented:
everything is green now.

Review comments gladly appreciated :)

gunchleoc commented:
> Can someone explain or point me to documentation about how replays are created/saved?

You'll have to dig through the code to find out, because I don't know either.

> What is GameController::GameType::kReplay actually used for? In the old code (and currently in this pr, too) it is not evaluated, is_valid_game always returns true, but I think this might be wrong in the first place?

It is used in the Lua scripting interface to switch off some functions that would crash replays. E.g.

```
int LuaPlayer::message_box(lua_State* L) {
	Game& game = get_game(L);
	// don't show message boxes in replays, cause they crash the game
	if (game.game_controller()->get_game_type() == GameController::GameType::kReplay) {
		return 1;
	}
```

Also used by the save handler - very important!

```
void SaveHandler::think(Widelands::Game& game) {
	if (!allow_saving_ || game.is_replay()) {
		return;
	}
```

gunchleoc reviewed (changes requested):
I have added some comments with suggestions. Not tested yet.

jmoerschbach committed:
added translations comments / return if table is empty to save if else if else

jmoerschbach committed:
Merge branch 'refactor-saveloading-ui' of https://github.com/jmoerschbach/widelands into refactor-saveloading-ui

jmoerschbach committed:
improve if else if else / move translator comment to correct place

jmoerschbach committed:
next try moving translation comment to right place...

jmoerschbach committed:
moved translation comments once again -> no warning locally...

gunchleoc reviewed (changes requested):
I have done some testing now, and I'd like to see 2 changes:

1. When saving a game from replay, it has `gametype="4"` and should be accessible from both singleplayer and multiplayer load screen. You can create one by loading a replay and then saving it. I guess that's what `GameController::GameType::kReplay` really is for.
2. When a directory is selected, show the directory name in the game details, just like we do with map details when starting a new game from map

For a follow-up branch, it would be great it the player could create new directories - you can find some code on the editor save map screen for that. The icing on the cake would be to be able to delete empty directories too.

jmoerschbach committed:
do not expose SavegameData#gametype anymore

jmoerschbach committed:
display directories in gamedetails

jmoerschbach committed:
Merge branch 'master' into refactor-saveloading-ui

jmoerschbach committed:
added todo

jmoerschbach commented:
Hi @gunchleoc 

sorry for the delayed response...

I implemented 2. Please have look I hope you like it :)

I'm not sure about 1. though. For me it is still quite confusing. For the time being, I would like to have it staying as it is - I am quite sure it is the exact same behaviour as current master. This branch was intended to introduce no different behaviour but only refactoring the code (which is not quite true anymore, since I implemented your wish number 2 ;))

However, I think I found another bug: Loading an unzipped replay crashes the game. Preloading, however, works fine (so it has nothing to do with the code I changed). I belive in current master, unzipped replays are not shown in the table at all, same in b20. Do you have any insights to this?

jmoerschbach committed:
added ()

frankystone commented:
> However, I think I found another bug: Loading an unzipped replay crashes the game. Preloading, however, works fine (so it has nothing to do with the code I changed). I belive in current master, unzipped replays are not shown in the table at all, same in b20. Do you have any insights to this?

Can't reproduce. Unzipped the file  .wrpl.wgf into a new folder with the same name and watching the replay works fine here. Did you unzipped it into a folder with the same name?

gunchleoc commented:
> > However, I think I found another bug: Loading an unzipped replay crashes the game. Preloading, however, works fine (so it has nothing to do with the code I changed). I belive in current master, unzipped replays are not shown in the table at all, same in b20. Do you have any insights to this?
> 
> Can't reproduce. Unzipped the file .wrpl.wgf into a new folder with the same name and watching the replay works fine here. Did you unzipped it into a folder with the same name?

It would be good to catch crashes like this in the future. However, we always natively zip the replays even if the option not to compress files has been chosen, so we have never really supported this. No need to fix it in this branch.

jmoerschbach commented:
> 
> 
> Never hard-code word order. Use "Selected %1% directory and %2%" and add a translators comment

I am not quite sure how to do that. Can you show me the complete example?
Actually, I think I want to enforce order here, because in the displayed list, directories are always shown first and then files, for example:
Selected 1 directory and 2 files:
- save/subfolder
- savegameOne
- savegameTwo

Selected 0 directories and 1 file:
- savegameOne

Selected 3 directories and 0 files:
- save/subfolderOne
- save/subfolderTwo
- save/subfolderThree

jmoerschbach committed:
added whitespaces

gunchleoc commented:
> > Never hard-code word order. Use "Selected %1% directory and %2%" and add a translators comment
> 
> I am not quite sure how to do that. Can you show me the complete example?
> Actually, I think I want to enforce order here, because in the displayed list, directories are always shown first and then files, for example:

I am a translator who translates into a language unlike English, so a big -1 on that. The only thing that you will be enforcing here is embarrassing grammar errors and convoluted linguistic contortions to try to make the translations work. It's one of the things that developers get wrong and that wants to make me bang my head against the wall.

jmoerschbach committed:
correct assembling of translatable strings with translator comments

jmoerschbach commented:
hope it is ok now :)

jmoerschbach committed:
added additional translater comment

Co-Authored-By: gunchleoc <fios@foramnagaidhlig.net>
